### PR TITLE
CLI to interact with Merkle Tree

### PR DESCRIPTION
**Pull request by JereSalo**
Implementation of CLI. 
Run the program with cargo run and then interact with the CLI using the tree command. Valid operations are shown in display. See the `readme` file for more information.

### CODE REVIEW

**Review by iinaki — CHANGES_REQUESTED** [review]
While I was testing the CLI I noticed there is no option to send an already hashed element, you could add a flag to the `add` command, something like `-h`, to indicate that the element doesn't need to be hashed. By doing this you could take more advantage of the `build_without_hashing` fn.

**Review by iinaki — CHANGES_REQUESTED** [review]
Other thing I noticed is that the tree allows the user to insert multiple times the same element, I think this shouldn't be allowed.

**Review by iinaki — CHANGES_REQUESTED** [review]
I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? 
I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way.
Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have.
Let me know what you think about this one.

File: src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            command: Commands::Show, // Placeholder, will be overridden by parsing
+            mktree: MerkleTree::new_empty(),
+        }
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        loop {
+            // Display prompt
+            print!("tree> ");
+            io::stdout().flush()?; // Flush prompt to the terminal
```
Comment (iinaki) [inline src/cli.rs:65]: For example here what I would do is the following:

```suggestion
            io::stdout().flush().map_err(|e| MerkleError::FailedToFlush(e.to_string()))?; // Flush prompt to the terminal
```

File: src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            // Execute the command and handle any errors
+            if let Err(e) = self.execute_command(command) {
+                println!("Error: {}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn execute_command(&mut self, command: Commands) -> Result<()> {
```
Comment (iinaki) [inline src/cli.rs]: Here I would do this:

```suggestion
    fn execute_command(&mut self, command: Commands) -> Result<(), MerkleError> {
```

**Review by iinaki — CHANGES_REQUESTED** [review]
Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using. 

When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type. 

By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.

**Comment by JereSalo** [comment]
> While I was testing the CLI I noticed there is no option to send an already hashed element, you could add a flag to the `add` command, something like `-h`, to indicate that the element doesn't need to be hashed. By doing this you could take more advantage of the `build_without_hashing` fn.

Done! I implemented it both for `build` and `add`, but using `-H` instead of `-h` because the latter is for help. 
Also I unified the build functions into one because I like to have those in the same place. Now it is just `build` and it receives a bool alongside the elements or hashes. The only problem that I see with what I did is that now I can't say the argument is either a hash or an element, because it can be both. But despite that, I like how it looks.

**Comment by JereSalo** [comment]
> Other thing I noticed is that the tree allows the user to insert multiple times the same element, I think this shouldn't be allowed.

Fixed! Both in `build` and `add`.

**Comment by JereSalo** [comment]
> I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way. Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have. Let me know what you think about this one.

Actually the CLI does use the `verify` function in the Merkle Tree. You send the `hash` you want to verify and it's `proof`. The CLI parses the proof into a `Vec<ProofElem>` and then calls the `verify` method in Merkle Tree with the given hash and the now parsed proof.

**Comment by JereSalo** [comment]
> Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using.
> 
> When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type.
> 
> By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.

I understand your point of view. I changed the way I handle errors and removed the anyhow crate but I didn't map IO errors to be MerkleErrors. I prefer to keep them as IO errors and leave the MerkleErrors just for the errors related to the Merkle Tree specifically. I hope we find a middle ground, but do tell me if you think I'm completely wrong.

File: src/cli.rs (hunk 2 above)
Comment (JereSalo) [reply src/cli.rs]: I agree. Done!

File: src/proof_element.rs [hunk 3]
```diff
@@ -1,13 +1,19 @@
+use crate::side::Side;
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct ProofElement {
     pub hash: String,
```
Comment (iinaki) [inline src/proof_element.rs]: Here hash can be of the Hash type right?

File: src/cli.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
+use crate::{
+    merkle_error::MerkleError, merkle_tree::MerkleTree, proof_element::ProofElement, side::Side,
+};
+// use anyhow::{Context, Error, Result};
```
Comment (iinaki) [inline src/cli.rs]: You left a comment here!

File: README.md [hunk 5]
```diff
@@ -1 +1,21 @@
...
+*Clone* the repository  
+*Enter* to the repository's folder  
+*Run* ```cargo run```  
+**Enjoy!**
+
+## Commands
+```tree build <elements> [--hashed]``` – Builds the tree with the given elements. Use the ```--hashed``` flag if the elements are already hashed.  
+```tree add <element> [--hashed]``` – Adds an element to the tree. Use the ```--hashed``` flag if the element is already hashed.  
+```tree show``` – Displays the tree structure on screen.  
+```tree verify <hash> <proof_file>``` – Verifies a if a proof is correct for a given hash in the tree.
```
Comment (iinaki) [inline README.md]:
```suggestion
```tree verify <hash> <proof_file>``` – Verifies if a proof is correct for a given hash in the tree.  
```

**Comment by iinaki** [comment]
> > I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way. Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have. Let me know what you think about this one.
> 
> Actually the CLI does use the `verify` function in the Merkle Tree. You send the `hash` you want to verify and it's `proof`. The CLI parses the proof into a `Vec<ProofElem>` and then calls the `verify` method in Merkle Tree with the given hash and the now parsed proof.

You're right I missed it, looks perfect to me!

**Comment by iinaki** [comment]
> > Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using.
> > When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type.
> > By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.
> 
> I understand your point of view. I changed the way I handle errors and removed the anyhow crate but I didn't map IO errors to be MerkleErrors. I prefer to keep them as IO errors and leave the MerkleErrors just for the errors related to the Merkle Tree specifically. I hope we find a middle ground, but do tell me if you think I'm completely wrong.

No, it looks very logical to me now! It's fine to separate them in this case actually because the two error kinds mean completely different things and should be managed differently. It wouldn't make sense to end the execution due to a command error from the user, but it does if there was an error with the i/o handling of the stdin, and this is what you are doing!

**Review by iinaki — APPROVED** [review]
Left a few comments, just typos, but other than that great work! I like how you handled the parsing of the commands in the CLI with `try_parse_from`! 🚀🚀🚀

File: src/proof_element.rs (hunk 3 above)
Comment (JereSalo) [reply src/proof_element.rs]: Done!

File: src/cli.rs (hunk 4 above)
Comment (JereSalo) [reply src/cli.rs]: Good catch!

File: README.md (hunk 5 above)
Comment (JereSalo) [reply README.md]: Done!

**Pull request merged**